### PR TITLE
[BE][PG NCCL] Improve input mismatch error msg

### PR DESCRIPTION
**Pull request by rohan-varma**
Test Plan: CI

Differential Revision: D48363238

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1502,8 +1502,16 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
...
-          " participants");
+          c10::str(
+              "Tensor list input to scatter/gather must match number of collective participants ",
+              "but got ",
+              tensor_lists[i].size(),
+              " inputs",
+              " with world_size",
+              world_size,
+              " and ",
+              num_devices,
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1513]: this is always 1 afaik, so can also exclude it.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -1502,8 +1502,16 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
...
       TORCH_CHECK(
           false,
-          "Tensor list input to scatter/gather must match number of collective"
-          " participants");
+          c10::str(
+              "Tensor list input to scatter/gather must match number of collective participants ",
+              "but got ",
+              tensor_lists[i].size(),
+              " inputs",
+              " with world_size",
```
Comment (awgu) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1510]: nit: I think?
```suggestion
              " with world_size ",
```

**Review by awgu — APPROVED** [review]
LGTM! Double check the nit.

**Review by H-Huang — APPROVED** [review]

**Review by fegin — APPROVED** [review]

**Pull request closed without merging**